---
title: "assignment"
output: html_document
date: "2025-03-16"
---

## Preparation steps

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(rtracklayer)
  library(epiwraps)
  library(GenomicRanges)
})
if (!requireNamespace("R.utils", quietly = TRUE)) {
  install.packages("R.utils")
}
library(R.utils)
library(ggplot2)
```

## Downloading the data

Downloading peak data from ENCODE for p300, H3K4me3, H3K4me1, H3K27ac and H3K27me3 in the mouse embryonic stem cell line Bruce 4 and unzipping it

```{r}
#download the peaks (always as bed narrowPeak)
options(timeout=3600) 


download.file("https://www.encodeproject.org/files/ENCFF460EOA/@@download/ENCFF460EOA.bed.gz", "P300_peaks.bed.gz") #conservative IDR threshold peaks
download.file("https://www.encodeproject.org/files/ENCFF247GVM/@@download/ENCFF247GVM.bed.gz", "H3K4me3_peaks.bed.gz") #replicated peaks
download.file("https://www.encodeproject.org/files/ENCFF824AFZ/@@download/ENCFF824AFZ.bed.gz", "H3K4me1_peaks.bed.gz") #pseudoreplicated peaks
download.file("https://www.encodeproject.org/files/ENCFF360VIS/@@download/ENCFF360VIS.bed.gz", "H3K27ac_peaks.bed.gz") #pseudoreplicated peaks
download.file("https://www.encodeproject.org/files/ENCFF105NKG/@@download/ENCFF105NKG.bed.gz", "H3K27me3_peaks.bed.gz")


# Unzip the BED files
gunzip("P300_peaks.bed.gz")
gunzip("H3K4me3_peaks.bed.gz")
gunzip("H3K4me1_peaks.bed.gz")
gunzip("H3K27ac_peaks.bed.gz")
gunzip("H3K27me3_peaks.bed.gz")
```

## Converting the BED-files to GRanges objects

1)  P300

```{r}
#Look at the BED-file
bed_file <- read.table("P300_peaks.bed", header = FALSE, nrows = 10)
print(bed_file)

```

```{r}
# FOR P300: Import BED files as GRanges objects

# Keep only the first three columns (chromosome, start, end) and assign appropriate names
p300_peaks <- read.table("P300_peaks.bed", header = FALSE)[, 1:3]
colnames(p300_peaks) <- c("chr", "start", "end")

# Convert to GRanges object
p300_peaks_gr <- GRanges(
  seqnames = p300_peaks$chr,
  ranges = IRanges(start = p300_peaks$start, end = p300_peaks$end)
)

# Inspect the GRanges object
p300_peaks_gr
```

2)  H3K4me3

```{r}
# Repeat for H3K4me3

#Look at the BED-file
bed_file <- read.table("H3K4me3_peaks.bed", header = FALSE, nrows = 10)
print(bed_file)
```

```{r}
# Read the BED file and keep only the first three columns (chromosome, start, end) and assign appropriate names
H3K4me3_peaks <- read.table("H3K4me3_peaks.bed", header = FALSE)[, 1:3]
colnames(H3K4me3_peaks) <- c("chr", "start", "end")

# Convert to GRanges object
H3K4me3_peaks_gr <- GRanges(
  seqnames = H3K4me3_peaks$chr,
  ranges = IRanges(start = H3K4me3_peaks$start, end = H3K4me3_peaks$end)
)

# Inspect the GRanges object
H3K4me3_peaks_gr
```

3)  H3K4me1

```{r}
# Repeat for H3K4me1

#Look at the BED-file
bed_file <- read.table("H3K4me1_peaks.bed", header = FALSE, nrows = 10)
print(bed_file)

```

```{r}
# Read the BED file and keep only the first three columns (chromosome, start, end) and assign appropriate names
H3K4me1_peaks <- read.table("H3K4me1_peaks.bed", header = FALSE)[, 1:3]
colnames(H3K4me1_peaks) <- c("chr", "start", "end")

# Convert to GRanges object
H3K4me1_peaks_gr <- GRanges(
  seqnames = H3K4me1_peaks$chr,
  ranges = IRanges(start = H3K4me1_peaks$start, end = H3K4me1_peaks$end)
)

# Inspect the GRanges object
H3K4me1_peaks_gr
```

4)  H3K27ac

```{r}
# Repeat for H3K27ac

#Look at the BED-file
bed_file <- read.table("H3K27ac_peaks.bed", header = FALSE, nrows = 10)
print(bed_file)
```

```{r}
# Read the BED file and keep only the first three columns (chromosome, start, end) and assign appropriate names
H3K27ac_peaks <- read.table("H3K27ac_peaks.bed", header = FALSE)[, 1:3]
colnames(H3K27ac_peaks) <- c("chr", "start", "end")

# Convert to GRanges object
H3K27ac_peaks_gr <- GRanges(
  seqnames = H3K27ac_peaks$chr,
  ranges = IRanges(start = H3K27ac_peaks$start, end = H3K27ac_peaks$end)
)

# Inspect the GRanges object
H3K27ac_peaks_gr
```

5)  H3K27me3

```{r}
# Repeat for H3K27me3

#Look at the BED-file
bed_file <- read.table("H3K27me3_peaks.bed", header = FALSE, nrows = 10)
print(bed_file)

```

```{r}
# Read the BED file and keep only the first three columns (chromosome, start, end) and assign appropriate names
H3K27me3_peaks <- read.table("H3K27me3_peaks.bed", header = FALSE)[, 1:3]
colnames(H3K27me3_peaks) <- c("chr", "start", "end")

# Convert to GRanges object
H3K27me3_peaks_gr <- GRanges(
  seqnames = H3K27me3_peaks$chr,
  ranges = IRanges(start = H3K27me3_peaks$start, end = H3K27me3_peaks$end)
)

# Inspect the GRanges object
H3K27me3_peaks_gr
```

## Finding the overlaps between p300 and the four Histone marks

1)  H3K4me3

```{r}
# Find overlaps between p300 and H3K4me3 
overlap_H3K4me3 <- findOverlaps(H3K4me3_peaks_gr, p300_peaks_gr) # to find all the overlapping regions, chose "any" as lose quite a lot with "within"

# Calculate percentage of p300 peaks that overlap with H3K4me3
percentage_H3K4me3 <- length(unique(subjectHits(overlap_H3K4me3))) / length(p300_peaks_gr)*100
percentage_H3K4me3
cat("Proportion of P300 peaks that overlap H3K4me3:", percentage_H3K4me3, "%\n")

```

2)  H3K4me1

```{r}
# Find overlaps between p300 and H3K4me1 
overlap_H3K4me1 <- findOverlaps(H3K4me1_peaks_gr, p300_peaks_gr)

# Calculate percentage of p300 peaks that overlap with H3K4me1
percentage_H3K4me1 <- length(unique(subjectHits(overlap_H3K4me1))) / length(p300_peaks_gr)*100
percentage_H3K4me1
cat("Proportion of P300 peaks that overlap H3K4me1:", percentage_H3K4me1, "%\n")
```

3)  H3K27ac

```{r}
# Find overlaps between p300 and H3K27ac
overlap_H3K27ac <- findOverlaps(H3K27ac_peaks_gr, p300_peaks_gr)

# Calculate percentage of p300 peaks that overlap with H3K27ac
percentage_H3K27ac <- length(unique(subjectHits(overlap_H3K27ac))) / length(p300_peaks_gr)*100
percentage_H3K27ac
cat("Proportion of P300 peaks that overlap H3K27ac:", percentage_H3K27ac, "%\n")
```

4)  H3K27me3

```{r}
# Find overlaps between p300 and H3K27me3
overlap_H3K27me3 <- findOverlaps(H3K27me3_peaks_gr, p300_peaks_gr)

# Calculate percentage of p300 peaks that overlap with H3K27me3
percentage_H3K27me3 <- length(unique(subjectHits(overlap_H3K27me3))) / length(p300_peaks_gr)*100
percentage_H3K27me3
cat("Proportion of P300 peaks that overlap H3K27me3:", percentage_H3K27me3, "%\n")
```

## Find more relevant overlaps

Alternatively, we could maybe try and exclude less relevant overlaps by setting a minimum overlap-length.

1)  p300 and H3K4me3

```{r}
# Find overlaps between p300 and H3K4me3 
hist(width(p300_peaks_gr))
```

```{r}
hist(width(H3K4me3_peaks_gr), breaks=75)
```

The two histograms show p300 has average length of 300bp and for H3K4me3 most reads are ca 200bp long, but up to 2000bp. We could specify the lengths of the overlaps to just include overlaps of a minimum of 50bp to get more meaningful overlaps.

```{r}
# Find overlaps between p300 and H3K4me3
overlap_H3K4me3_min <- findOverlaps(H3K4me3_peaks_gr, p300_peaks_gr, minoverlap = 50) # here only regions overlapping by a minimum of 50bp are reported to hopefully exclude less relevant overlaps, but not lose any as histogram shows high proportion of reads of small length

# Calculate percentage of p300 peaks that overlap with H3K4me3
percentage_H3K4me3_min <- length(unique(subjectHits(overlap_H3K4me3_min))) / length(p300_peaks_gr)*100
percentage_H3K4me3_min
cat("Proportion of P300 peaks that overlap H3K4me3:", percentage_H3K4me3_min, "%\n")
```

2)  p300 and H3K4me1

```{r}
hist(width(H3K4me1_peaks_gr), breaks=75) # here there are lots of small reads so will set minimum for very low number
```

```{r}
# Find overlaps between p300 and H3K4me1 
overlap_H3K4me1_min <- findOverlaps(H3K4me1_peaks_gr, p300_peaks_gr, minoverlap = 5)

# Calculate percentage of p300 peaks that overlap with H3K4me1
percentage_H3K4me1_min <- length(unique(subjectHits(overlap_H3K4me1_min))) / length(p300_peaks_gr)*100
percentage_H3K4me1_min
cat("Proportion of P300 peaks that overlap H3K4me1:", percentage_H3K4me1_min, "%\n")
```

2)  p300 and H3K27ac

```{r}
hist(width(H3K27ac_peaks_gr), breaks=75) # there are many peaks around 300, so we can maybe set the minimum a bit higher again to 50
```

```{r}
# Find overlaps between p300 and H3K4me1 
overlap_H3K27ac_min <- findOverlaps(H3K27ac_peaks_gr, p300_peaks_gr, minoverlap = 50)

# Calculate percentage of p300 peaks that overlap with H3K4me1
percentage_H3K27ac_min <- length(unique(subjectHits(overlap_H3K27ac_min))) / length(p300_peaks_gr)*100
percentage_H3K27ac_min
cat("Proportion of P300 peaks that overlap H3K27ac:", percentage_H3K27ac_min, "%\n")

```

3)  p300 and H3K27me3

```{r}
hist(width(H3K27me3_peaks_gr), breaks=75) # many peaks around 200, so maybe we can set the minimum for 50 again
```

```{r}
# Find overlaps between p300 and H3K27me3 
overlap_H3K27me3_min <- findOverlaps(H3K27me3_peaks_gr, p300_peaks_gr, minoverlap = 50)

# Calculate percentage of p300 peaks that overlap with H3K4me1
percentage_H3K27me3_min <- length(unique(subjectHits(overlap_H3K27me3_min))) / length(p300_peaks_gr)*100
percentage_H3K27me3_min
cat("Proportion of P300 peaks that overlap H3K27me3:", percentage_H3K27me3_min, "%\n")

```
