---
title: "Assignment"
output: html_document
date: "2025-04-02"
---

# Installing all packages

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  #library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives 
  library(motifmatchr) # for scanning sequences for matches of given motifs
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs -> contains a collection of DNA-binding motifs, not fully up to date
  library(TFBSTools) # for handling some motif formats 
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis -> performs the motif enrichment analysis
})
if (!requireNamespace("AnnotationHub", quietly = TRUE)) {
    BiocManager::install("AnnotationHub")
}
library(AnnotationHub)
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Install Bioconductor packages if not already installed
BiocManager::install(c("AnnotationHub", "GenomicRanges", "MotifDb", 
                       "Biostrings", "rtracklayer", "motifmatchr"))

library(rtracklayer)     # Reading/writing genomic files
BiocManager::install("BSgenome.Mmusculus.UCSC.mm10")
library(BSgenome.Mmusculus.UCSC.mm10)
```

# Downloading mus musculus peaks for transcription factor "REST" in myocytes

```{r}
download.file("https://www.encodeproject.org/files/ENCFF368VWJ/@@download/ENCFF368VWJ.bed.gz", "mMyo_REST.bed.gz")
peaks <- rtracklayer::import("mMyo_REST.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "UCSC" # downloading in USCS-format to match BSgenome format
head(peaks)
```

# Identify instances of factor's motifs

## a) In peaks

1)  Downloading mus musculus reference genome

```{r}
ah <- AnnotationHub()
AnnotationHub::query(ah, c("Mus Musculus", "2bit")) # find AH14005 as most recent
```

```{r}
genome <- ah[["AH14005"]]
```

```{r}
genome_seqs <- import(genome)
```

2)  Search for "REST" in motif database

```{r}
motifs <- query(MotifDb, "REST")
names(motifs)
```

```{r}
motif <- motifs[["Mmusculus-HOCOMOCOv10-REST_MOUSE.H10MO.A"]] # select Mus musculus HOCOMOCO
motif
```

```{r}
view_motifs(motifs[1:2]) # visualize motif
```

```{r}
Biostrings::writeXStringSet(genome_seqs, "genome.fa") # convert genome to fasta format
motif2 <- convert_motifs(motif, class="TFBSTools-PWMatrix") # convert motif to format accepted by package
```

```{r}
moi <- motifmatchr::matchMotifs(motif2, subject=peaks, genome= BSgenome.Mmusculus.UCSC.mm10,
                                out="positions") # returns genomic positions of each motif match
moi <- moi[[1]] # get results of first motif
head(moi)
```

5)  How many of the peaks have a motif?

```{r}
table(overlapsAny(peaks, moi))
```

```{r}
(1657/3555)*100
```

Of the 3555 peaks, 1657 (46.6%) contain a motif.

## b) In whole genome

1)  Scan whole genome

```{r}
motif_across_genome <- matchMotifs(motif2, subject=genome_seqs, out="positions")

motif_across_genome <- motif_across_genome[[1]]
names(motif_across_genome) <- names(genome_seqs)
motif_across_genome <- as(motif_across_genome, "GRanges")
head(motif_across_genome)
```

```{r}
seqlevelsStyle(motif_across_genome) <- "UCSC"
table(overlapsAny(motif_across_genome, peaks)) 
```

```{r}
(2253/601816)*100
```

Of the 601816 motif instances, 2253 (0.37%) overlap a peak.
